src/service/position-observer/position-observer-impl.js

```diff
@@ -0,0 +1,303 @@
+/**
+ * Copyright 2017 The AMP HTML Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use baseInstance file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {registerServiceBuilderForDoc} from '../../service';
+import {Services} from '../../services';
+import {dev} from '../../log';
+import {layoutRectLtwh} from '../../layout-rect';
+import {parseJson} from '../../json.js';
+import {getData} from '../../event-helper';
+import {debounce} from '../../utils/rate-limit';
+import {MessageType} from '../../3p-frame-messaging';
+import {PositionObserverEntry} from './position-observer-entry';
+import {
+  PosObViewportInfoAmpDoc,
+} from './position-observer-viewport-info';
+import {
+  PositionObserverFidelity,
+  LOW_FIDELITY_FRAME_COUNT,
+} from './position-observer-fidelity';
+/** @const @private */
+const TAG = 'POSITION_OBSERVER';
+
+
+export class PositionObserver {
+  /**
+   * @param {!Window} win
+   * @param {!../vsync-impl.Vsync} vsync
+   * @param {!./position-observer-viewport-info.PosObViewportInfoDef} posObViewportInfo
+   */
+  constructor(win, vsync, posObViewportInfo) {
+    /** @private {!Window} */
+    this.win_ = win;
+
+    /** @private {!Array<!Object>} */
+    this.entries_ = [];
+
+    /** @private {!../vsync-impl.Vsync} */
+    this.vsync_ = vsync;
+
+    /** @private {!./position-observer-viewport-info.PosObViewportInfoDef} */
+    this.viewport_ = posObViewportInfo;
+
+    /** @private {Array<function()>} */
+    this.unlisteners_ = [];
+
+    /** @private {boolean} */
+    this.inScroll_ = false;
+
+    /** @private {boolean} */
+    this.measure_ = false;
+
+    /** @private {function()} */
+    this.boundStopScroll_ = debounce(this.win_, () => {
+      this.inScroll_ = false;
+    }, 500);
+
+    /** @private {boolean} */
+    this.needRefreshOnMessage_ = false;
+  }
+
+  /**
+   * @param {!Element} element
+   * @param {PositionObserverFidelity} fidelity
+   * @param {function(?./position-observer-entry.PositionInViewportEntryDef)} handler
+   */
+  observe(element, fidelity, handler) {
+    const entry = new PositionObserverEntry(element, fidelity, handler);
+
+    this.entries_.push(entry);
+
+    if (this.entries_.length == 1) {
+      this.startCallback_();
+    }
+
+    this.vsync_.measure(() => {
+      this.updateSingleEntry_(entry);
+    });
+  }
+
+  /**
+   * @param {!Element} element
+   */
+  unobserve(element) {
+    for (let i = 0; i < this.entries_.length; i++) {
```

Comment:
Not exactly. `filterSplice` splice every array item that filter false. Here we can use some optimization to only splice the first item that filter false. 
Do you think it is worth to create a new function for this?